iree/samples/variables_and_state/main.c

```diff
@@ -0,0 +1,195 @@
+  IREE_RETURN_IF_ERROR(iree_hal_buffer_view_clone_heap_buffer(
+      iree_runtime_session_device_allocator(call_add_to_value->session),
+      arg0_shape, IREE_ARRAYSIZE(arg0_shape), IREE_HAL_ELEMENT_TYPE_SINT_32,
+      IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+      IREE_HAL_BUFFER_USAGE_ALL,
+      iree_make_const_byte_span((void*)arg0_data, sizeof(arg0_data)), &arg0));
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_inputs_push_back_buffer_view(call_add_to_value, arg0));
+  iree_hal_buffer_view_release(arg0);
+  IREE_RETURN_IF_ERROR(
+      iree_runtime_call_invoke(call_add_to_value, /*flags=*/0));
+  return iree_ok_status();
+}
+
+iree_status_t counter_reset_value(iree_runtime_call_t* call_reset_value) {
+  return iree_runtime_call_invoke(call_reset_value, /*flags=*/0);
+}
+
+iree_status_t run_sample(iree_string_view_t bytecode_module_path,
+                         iree_string_view_t driver_name) {
+  //===-------------------------------------------------------------------===//
+  // Instance configuration (this should be shared across sessions).
+  fprintf(stdout, "Configuring IREE runtime instance and '%s' device\n",
+          driver_name.data);
+  iree_runtime_instance_options_t instance_options;
+  iree_runtime_instance_options_initialize(IREE_API_VERSION_LATEST,
+                                           &instance_options);
+  iree_runtime_instance_options_use_all_available_drivers(&instance_options);
+  iree_runtime_instance_t* instance = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_instance_create(
+      &instance_options, iree_allocator_system(), &instance));
+  // TODO(#5724): move device selection into the compiled modules.
+  iree_hal_device_t* device = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_instance_try_create_default_device(
+      instance, driver_name, &device));
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Session configuration (one per loaded module to hold module state).
+  fprintf(stdout, "Creating IREE runtime session\n");
+  iree_runtime_session_options_t session_options;
+  iree_runtime_session_options_initialize(&session_options);
+  iree_runtime_session_t* session = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_session_create_with_device(
+      instance, &session_options, device,
+      iree_runtime_instance_host_allocator(instance), &session));
+  iree_hal_device_release(device);
+
+  fprintf(stdout, "Loading bytecode module at '%s'\n",
+          bytecode_module_path.data);
+  IREE_RETURN_IF_ERROR(iree_runtime_session_append_bytecode_module_from_file(
+      session, bytecode_module_path.data));
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Look up functions.
+  iree_runtime_call_t call_get_value;
```

Comment:
it may be worth seeing as we add the next example, but the caching of the calls is strictly an optimization and not required - now that we have the nice functions it may be nicer to not cache these and just init/deinit them inside of the counter_x() methods for clarity - would simplify the code flow here and avoid confusion around the lifetime of things.

Reply:
Dropped the caching. This also makes it easier to follow the error handling and lifetimes. 

I'm now also noticing that the main function leaks on errors:
```c
  IREE_RETURN_IF_ERROR(...
  IREE_RETURN_IF_ERROR(...

  ...

  iree_runtime_session_release(session);
  iree_runtime_instance_release(instance);
```

Should I change those to the 
```c
if (iree_status_is_ok(status)) {
    status = ...
}
```
style as well?

Reply:
Made that change (to status handling) as well.